{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a8c3cc2-08a9-436d-9d0a-3f83ba613975",
   "metadata": {},
   "source": [
    "# Meta'omics for Ocean Science\n",
    "\n",
    "Ocean Hack Week 2023 Tutorial by Julia M Brown\n",
    "\n",
    "![scales](./tutorial_images/rainbow_satellite_to_microbe.png)  \n",
    "\n",
    "With thanks to the following for content and inspiration:  \n",
    "[Greg Gavelis](https://github.com/ggavelis), [Joe Brown](https://github.com/brwnj), [Maria Pachiadaki](https://github.com/microbiaki), [Ramunas Stepanauskas](https://www.bigelow.org/about/people/rstepanauskas.html), [MerenLab](https://merenlab.org/), [Kaiju Team](https://bioinformatics-centre.github.io/kaiju/), [Cath Mitchell](https://github.com/MarineOpticsLab)\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a33a281-5db8-4d59-867b-de2d403fcc34",
   "metadata": {},
   "source": [
    "### What is 'omics data?\n",
    "\n",
    "* Data on biological molecules  \n",
    "* 'Meta' refers to collecting and processing samples in bulk \n",
    "* Data often focused on specific size fractions  \n",
    "\n",
    "### How is it generated?\n",
    "\n",
    "* Collection of sample in bulk.  \n",
    "* For planktonic microbes, samples are collected based on a specific size fraction that targets different microbial groups\n",
    "    * e.g. bacteria and archaea, protists, phytoplankton, viruses  \n",
    "* Nucleic acids, proteins or other target molecules extracted and sequenced \n",
    "* For nucleotide data (DNA + RNA), samples often sequenced via Illumina sequencing\n",
    "    * generates short __paired end__ reads \n",
    "    * reads can be characterized directly or used to assemble larger __contiguous sequences__\n",
    "\n",
    "![omics](./tutorial_images/metaXomics_diagram.png)  \n",
    "\n",
    "### It can tell us the who and what of microbial communities\n",
    "\n",
    "**Metagenome (DNA)** : Presence and potential  \n",
    "**Metatranscriptome (RNA)**: Activity  \n",
    "\n",
    "**Taxonomy:**  \n",
    "\n",
    "* What microbes are present -- DNA\n",
    "* Which microbes are active -- RNA\n",
    "\n",
    "**Function:**\n",
    "\n",
    "* What is the metabolic potential? -- DNA\n",
    "* What processes are being carried out? -- RNA\n",
    "\n",
    "### What does it look like?  \n",
    "\n",
    "* fastq - raw sequence read data with quality information included\n",
    "* fasta - sequence data, can be contiguous sequences, open reading frames (i.e. coding sequences) or protein sequences\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0fe7e5-b7ae-4c6f-a11c-45c18cfb7ab5",
   "metadata": {},
   "source": [
    "### How can we use raw reads?\n",
    "\n",
    "**Read profiling** is one of the most commonly used processes in 'omics analysis. It is applied to access the relative abundance of taxonomic groups within metagenomic datasets (when using DNA metagenomes) or to estimate the expression of different microbial taxa (when RNA metatranscriptomes are used).\n",
    "\n",
    "In a nutshell short reads are aligned to a genomic reference sequences, which have taxonomic information assigned to them that may be assigned to the reads.\n",
    "\n",
    "![recruitment](./tutorial_images/01-metagenomic-read-recruitment-simple.gif)  \n",
    "(Thank you [MerenLab](https://merenlab.org/) for the animation)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86297889-8a11-4b44-940b-a1d50c54107c",
   "metadata": {},
   "source": [
    "# Read classification tools\n",
    "\n",
    "# [Kaiju](https://kaiju.binf.ku.dk/server)\n",
    "![Kaiju](https://kaiju.binf.ku.dk/images/kaiju3_header.gif)\n",
    "\n",
    "**Also: [Kraken2](https://github.com/DerrickWood/kraken2/)**\n",
    "\n",
    "These workflows are wicked fast!\n",
    "\n",
    "### How do they work?\n",
    "\n",
    "**Database**  \n",
    "Database consists of a collection of translated proteins mapped to microbial genomes.\n",
    "\n",
    "<img src=\"./tutorial_images/proteins_to_genomes.png\" alt=\"p2g\" width=\"400\"/>\n",
    "\n",
    "### Short read alignment\n",
    "\n",
    "Reads are translated into protein sequences and aligned to reference protein sequences. Best matches to proteins are then taxonomically assigned based on protein's membership in microbial genomes.\n",
    "\n",
    "![kaiju_diagram](./tutorial_images/short_read_align.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06b27e46-297a-49c7-825c-35f6d74d5862",
   "metadata": {},
   "source": [
    "## Tools are as good as your reference database\n",
    "\n",
    "Kaiju and other classifiers rely on genome databases that primarily contain genomes from isolated microbes and genomes assembled from metagenomes ('MAGs').\n",
    "\n",
    "**Available Standard Kaiju Databases**  \n",
    "<img src=\"https://upload.wikimedia.org/wikipedia/commons/0/07/US-NLM-NCBI-Logo.svg\" alt=\"ncbi\" width=\"100\"/>\n",
    "\n",
    "**nr**: Non-redundant proteins from bacteria, archaea and viruses  \n",
    "**RefSeq**: Curated bacterial, archaeal and viral genomes from NCBI\n",
    "\n",
    "<img src=\"https://progenomes.embl.de/img/progene_head21.png\" alt=\"progenomes\" width=\"300\"/>  \n",
    "\n",
    "**ProGenomes**: Database of microbial genomes including MAGs from diverse environments\n",
    "\n",
    "**Note:** Kaiju has other available databases that could be useful for your environment or organisms of interest. See their website for more options.  \n",
    "\n",
    "**Database Limitations**: Despite the depth of these collections, they leave stones unturned. Microbial genomic diversity is high in marine systems and genomes assembled from short reads represent only a fraction of the microbial diversity present in the ocean!\n",
    "\n",
    "## SAGs\n",
    "\n",
    "Single Cell Genomics is another type of 'omics data that is well suited for reference genomic data. It consists of DNA sequence data generated from the DNA present in single cells.  Each set of data from each cell is referred to as a __Single Amplified Genome (i.e. SAG)__. SAGs represent real biological units recovered from samples, and contain genomic information specific to individuals.\n",
    "\n",
    "<img src=\"./tutorial_images/scg_diagram.png\" alt=\"sag\" width=\"600\"/>\n",
    "\n",
    "## GORG-Tropics: A collection of reference genomes from individual cells from the Tropical and Sub-tropical Epipelagic Ocean\n",
    "\n",
    "GORG-Tropics is more representative of global ocean microbes than MAGs or currently available reference genomes*.\n",
    "\n",
    "*I am not sure if GORG-Tropics has been integrated into ProGenomes or not\n",
    "\n",
    "![GORG-Figure2](https://ars.els-cdn.com/content/image/1-s2.0-S0092867419312735-gr2.jpg)  \n",
    "\n",
    "\n",
    "GORG-Tropics is more accurate and sensitive than default databases used for read classification by Kaiju when analyzing marine epipelagic samples. When GORG-Tropics used as a database for reads from similar environments, many more were able to be correctly classified. \n",
    "\n",
    "![GORG-Figure](https://ars.els-cdn.com/content/image/1-s2.0-S0092867419312735-gr6.jpg)  \n",
    "\n",
    "The other advantage of using a tailored database is that it takes up less storage space :)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0a86db9-7534-4fe9-aa48-440bd76afa7a",
   "metadata": {},
   "source": [
    "## TODAY\n",
    "\n",
    "We will be running Kaiju on a collection of epipelagic metagenomes from the Bermuda Atlantic Time Series using version 1 of the GORG-Tropics database."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d9e6f14-d44f-4abd-8c39-987bcbe1c361",
   "metadata": {},
   "source": [
    "## Data Prep\n",
    "\n",
    "For this lesson, we look at microbial community dynamics over time at BATs using short read alignment with Kaiju against the GORG-Tropics database.\n",
    "\n",
    "The metagenomes we will use are a small subset of metagenomes reported in [this](https://www.nature.com/articles/sdata2018176) publication. These metagenomes are available through NCBI project ID [PRJNA385855](https://www.ncbi.nlm.nih.gov/bioproject?term=PRJNA385855). I've downloaded a metadata sheet with all sra metagenomes from this bioproject to: ./data/PRJNA385855_sra_metadata.csv\n",
    "\n",
    "Let's check this table out, and I'll show you which metagenomes I selected."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "25776333-d33e-47be-8ad3-f97952ad5fc3",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>bottle_id</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>...</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>Library Name</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SRA Study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR6507277</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>16133582100</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390922</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>2140200308</td>\n",
       "      <td>6618578156</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>22.75 N 158 W</td>\n",
       "      <td>S0627</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0627</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR6507278</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>15874959000</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390923</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>2160200304</td>\n",
       "      <td>6562862443</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>22.75 N 158 W</td>\n",
       "      <td>S0628</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0628</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR6507279</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>15069825300</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390924</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1024800503</td>\n",
       "      <td>6265839401</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>31.66 N 64.16 W</td>\n",
       "      <td>S0629</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0629</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR6507280</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>25807308000</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390925</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1025200510</td>\n",
       "      <td>10523504402</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>31.66 N 64.16 W</td>\n",
       "      <td>S0630</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0630</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR5720219</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>6713331000</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN07137016</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1640201117</td>\n",
       "      <td>2811014041</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>22.75 N 158 W</td>\n",
       "      <td>S0519</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0519</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run Assay Type  AvgSpotLen        Bases   BioProject     BioSample  \\\n",
       "0  SRR6507277        WGS         300  16133582100  PRJNA385855  SAMN08390922   \n",
       "1  SRR6507278        WGS         300  15874959000  PRJNA385855  SAMN08390923   \n",
       "2  SRR6507279        WGS         300  15069825300  PRJNA385855  SAMN08390924   \n",
       "3  SRR6507280        WGS         300  25807308000  PRJNA385855  SAMN08390925   \n",
       "4  SRR5720219        WGS         300   6713331000  PRJNA385855  SAMN07137016   \n",
       "\n",
       "                    BioSampleModel   bottle_id        Bytes Center Name  ...  \\\n",
       "0  MIMS.me,MIGS/MIMS/MIMARKS.water  2140200308   6618578156         MIT  ...   \n",
       "1  MIMS.me,MIGS/MIMS/MIMARKS.water  2160200304   6562862443         MIT  ...   \n",
       "2  MIMS.me,MIGS/MIMS/MIMARKS.water  1024800503   6265839401         MIT  ...   \n",
       "3  MIMS.me,MIGS/MIMS/MIMARKS.water  1025200510  10523504402         MIT  ...   \n",
       "4  MIMS.me,MIGS/MIMS/MIMARKS.water  1640201117   2811014041         MIT  ...   \n",
       "\n",
       "           lat_lon Library Name LibraryLayout LibrarySelection LibrarySource  \\\n",
       "0    22.75 N 158 W        S0627        PAIRED           RANDOM   METAGENOMIC   \n",
       "1    22.75 N 158 W        S0628        PAIRED           RANDOM   METAGENOMIC   \n",
       "2  31.66 N 64.16 W        S0629        PAIRED           RANDOM   METAGENOMIC   \n",
       "3  31.66 N 64.16 W        S0630        PAIRED           RANDOM   METAGENOMIC   \n",
       "4    22.75 N 158 W        S0519        PAIRED           RANDOM   METAGENOMIC   \n",
       "\n",
       "            Organism  Platform           ReleaseDate Sample Name  SRA Study  \n",
       "0  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0627  SRP109831  \n",
       "1  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0628  SRP109831  \n",
       "2  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0629  SRP109831  \n",
       "3  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0630  SRP109831  \n",
       "4  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0519  SRP109831  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"./data/PRJNA385855_sra_metadata.csv\", sep = \",\")\n",
    "\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1c83c861-0960-46fd-a4b0-7aedda241e7d",
   "metadata": {},
   "source": [
    "There's a lot of information here, for now, we just need the 'Run' ID so that we can use it to grab data from NCBI, as well as the collection date and depth."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "b0da40e5-3771-4593-bb43-da145f615afe",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 21 metagenomes\n"
     ]
    }
   ],
   "source": [
    "# selecting surface samples collected from BATS\n",
    "mgoi = df[df['cruise_id'].str.contains('BATS') & df['Depth'].isin(['10m','1m'])][['Run','Collection_date','cruise_id','BioSample','Depth']].sort_values(by = 'Collection_date')\n",
    "\n",
    "print('There are', len(mgoi), 'metagenomes')\n",
    "      \n",
    "# going to save this table to file\n",
    "mgoi.to_csv(\"./data/bats_metagenomes_of_interest.csv\", index=False)\n",
    "\n",
    "# I also saved a text file with one metagenome ID per line for downloading these metagenomes with a shell script\n",
    "with open('data/metagenomes_to_download.txt', 'w') as oh:\n",
    "    for run in mgoi['Run']:\n",
    "        print(run, file = oh)\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa9ca13a-de4e-420e-a3ba-4e0abdc26493",
   "metadata": {
    "tags": []
   },
   "source": [
    "I next downloaded these metagenomes from NCBI using a package called [fastq-dl](). These metagenomes are large, and downloading them takes a while, so I did this overnight using a small shell script:\n",
    "\n",
    "```\n",
    "cd ./data/\n",
    "\n",
    "while read p; \n",
    "    do fastq-dl -a $p --provider sra; \n",
    "done < metagenomes_to_download.txt\n",
    "```\n",
    "\n",
    "Next, for the purposes of this tutorial, I subsampled these paired metagenomes down to 100000k reads per metagenome using a package called [seqtk](). I was lazy and ran this in a jupyter notebook. The script I used to do this was:\n",
    "\n",
    "```\n",
    "%%python\n",
    "import glob\n",
    "\n",
    "ffqs = glob.glob(\"*_1.fastq.gz\")\n",
    "\n",
    "for f in ffqs:\n",
    "    r = f.replace('_1','_2')\n",
    "    \n",
    "    outf = \"./subsampled_metagenomes/{}\".format(f.replace(\"_1.fastq.gz\",\n",
    "                                                          '_100ksub_1.fastq.gz'))\n",
    "    outr = \"./subsampled_metagenomes/{}\".format(r.replace(\"_2.fastq.gz\",\n",
    "                                                          '_100ksub_2.fastq.gz'))\n",
    "    \n",
    "    !seqtk sample -s 123 {f} 100000 > {outf}\n",
    "    \n",
    "    !seqtk sample -s 123 {r} 100000 > {outr}\n",
    "```\n",
    "\n",
    "The important command is ```seqtk sample -s 123 {infastq} 100000 > {outfastq}```.  The script randomly subsamples a subset of reads. I needed to do this for paired metagenomic data, so to make sure the samples were random but consistent between the paired files I used the ```-s``` parameter to set the seed for random sampling. After the ```-s``` parameter (which can be any integer), I provide the input fastq file and then the number of reads I would like to subsample from the input file, I then send the output to a new file. \n",
    "\n",
    "This step also takes a while to run, so we will be not doing this in real time. The output files from this step can be found at: XXXXXXX"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19c26025-15a1-47bc-bfe1-06277f3c492e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Running Kaiju\n",
    "\n",
    "It's finally time!  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c2278fb-75a8-43e5-9063-9d52faabfd1c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
