{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "3a8c3cc2-08a9-436d-9d0a-3f83ba613975",
   "metadata": {},
   "source": [
    "# Meta'omics for Ocean Science\n",
    "\n",
    "![scales](./tutorial_images/rainbow_satellite_to_microbe.png)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a33a281-5db8-4d59-867b-de2d403fcc34",
   "metadata": {},
   "source": [
    "### What is 'omics data?\n",
    "\n",
    "* Data on biological molecules  \n",
    "* 'Meta' refers to collecting and processing samples in bulk. \n",
    "* Data often focused on specific size fractions  \n",
    "\n",
    "\n",
    "### What can it tell us?\n",
    "\n",
    "#### Taxonomy: \n",
    "* What microbes are present -- DNA\n",
    "* Which microbes are active -- RNA\n",
    "\n",
    "#### Function:\n",
    "* What is the metabolic potential? -- DNA\n",
    "* What processes are being carried out? -- RNA\n",
    "\n",
    "### How is it generated?\n",
    "![omics](./tutorial_images/metaXomics_diagram.png)  \n",
    "\n",
    "Data is generated through collecting bulk community samples. For planktonic microbes, samples are collected based on a specific planktonic size fraction that targets different microbial groups (e.g. bacteria and archaea, protists, phytoplankton, viruses, etc).Nucleic acids, proteins or other target molecules are then extracted and sequenced. For nucleotide data (DNA + RNA), samples are often sequenced via Illumina sequencing, generating short __paired end__ reads that can be characterized directly or used to assemble larger __contiguous sequences__. \n",
    "\n",
    "### What does it look like?\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0fe7e5-b7ae-4c6f-a11c-45c18cfb7ab5",
   "metadata": {},
   "source": [
    "### How can we use it?\n",
    "\n",
    "Read profiling is one of the most commonly used processes in 'omics analysis. It is applied to access the relative abundance of taxonomic groups within metagenomic datasets (when using DNA metagenomes) or to estimate the expression of different microbial taxa (when RNA metatranscriptomes are used).\n",
    "\n",
    "In a nutshell short reads are aligned to a genomic reference sequences, which have taxonomic information assigned to them that may be assigned to the reads.\n",
    "\n",
    "![recruitment](./tutorial_images/01-metagenomic-read-recruitment-simple.gif)  \n",
    "(Thank you [MerenLab](https://merenlab.org/) for the animation)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "86297889-8a11-4b44-940b-a1d50c54107c",
   "metadata": {},
   "source": [
    "# Read classification tools\n",
    "\n",
    "# [Kaiju](https://kaiju.binf.ku.dk/server)\n",
    "![Kaiju](https://kaiju.binf.ku.dk/images/kaiju3_header.gif)\n",
    "\n",
    "#### Alternative: [Kraken2](https://github.com/DerrickWood/kraken2/)\n",
    "\n",
    "### How do they work?\n",
    "Translated Reads --> Proteins --> Genomes\n",
    "\n",
    "![kaiju_diagram](./tutorial_images/read_classification_workflow.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "06b27e46-297a-49c7-825c-35f6d74d5862",
   "metadata": {},
   "source": [
    "# Tools are as good as your reference database\n",
    "\n",
    "Kaiju and other classifiers rely on genome databases that primarily contain genomes from isolated microbes and genomes assembled from metagenomes ('MAGs').\n",
    "\n",
    "__Isolates__ not representative of the microbial diversity of the planet, particularly in oligotrophic and extreme environments.  \n",
    "\n",
    "__MAGs__ also not representative of overal microbial diversity. For example, it is challenging to assemble a MAG of SAR11 (i.e. Pelagibacter) because their genomes are variable between individuals.\n",
    "\n",
    "\n",
    "# GORG-Tropics: A collection of reference genomes from individual cells from the Tropical and Sub-tropical Epipelagic Ocean\n",
    "![GORG-Figure2](https://ars.els-cdn.com/content/image/1-s2.0-S0092867419312735-gr2.jpg)  \n",
    "\n",
    "GORG-Tropics is more representative of global ocean microbes than MAGs.\n",
    "\n",
    "\n",
    "![GORG-Figure](https://ars.els-cdn.com/content/image/1-s2.0-S0092867419312735-gr6.jpg)  \n",
    "\n",
    "GORG-Tropics is more accurate and sensitive than default databases used for read classification by Kaiju. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f096e33c-dea8-4349-a0fc-20e0f268718b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "df = pd.read_csv(\"./data/PRJNA385855_sra_metadata.csv\", sep = \",\")\n",
    "mgoi = df[df['cruise_id'].str.contains('BATS') & df['Depth'].isin(['10m','1m'])][['Run','Collection_date','cruise_id','BioSample','Depth']].sort_values(by = 'Collection_date')\n",
    "\n",
    "\n",
    "\n",
    "# going to save this table to file\n",
    "mgoi.to_csv(\"./data/bats_metagenomes_of_interest.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c6ad4635-7f15-445f-8c5c-da2ce325d2bd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>BioSampleModel</th>\n",
       "      <th>bottle_id</th>\n",
       "      <th>Bytes</th>\n",
       "      <th>Center Name</th>\n",
       "      <th>...</th>\n",
       "      <th>lat_lon</th>\n",
       "      <th>Library Name</th>\n",
       "      <th>LibraryLayout</th>\n",
       "      <th>LibrarySelection</th>\n",
       "      <th>LibrarySource</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Platform</th>\n",
       "      <th>ReleaseDate</th>\n",
       "      <th>Sample Name</th>\n",
       "      <th>SRA Study</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SRR6507277</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>16133582100</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390922</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>2140200308</td>\n",
       "      <td>6618578156</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>22.75 N 158 W</td>\n",
       "      <td>S0627</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0627</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SRR6507278</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>15874959000</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390923</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>2160200304</td>\n",
       "      <td>6562862443</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>22.75 N 158 W</td>\n",
       "      <td>S0628</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0628</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SRR6507279</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>15069825300</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390924</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1024800503</td>\n",
       "      <td>6265839401</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>31.66 N 64.16 W</td>\n",
       "      <td>S0629</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0629</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SRR6507280</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>25807308000</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN08390925</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1025200510</td>\n",
       "      <td>10523504402</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>31.66 N 64.16 W</td>\n",
       "      <td>S0630</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0630</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SRR5720219</td>\n",
       "      <td>WGS</td>\n",
       "      <td>300</td>\n",
       "      <td>6713331000</td>\n",
       "      <td>PRJNA385855</td>\n",
       "      <td>SAMN07137016</td>\n",
       "      <td>MIMS.me,MIGS/MIMS/MIMARKS.water</td>\n",
       "      <td>1640201117</td>\n",
       "      <td>2811014041</td>\n",
       "      <td>MIT</td>\n",
       "      <td>...</td>\n",
       "      <td>22.75 N 158 W</td>\n",
       "      <td>S0519</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>RANDOM</td>\n",
       "      <td>METAGENOMIC</td>\n",
       "      <td>marine metagenome</td>\n",
       "      <td>ILLUMINA</td>\n",
       "      <td>2018-05-01T00:00:00Z</td>\n",
       "      <td>S0519</td>\n",
       "      <td>SRP109831</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 36 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          Run Assay Type  AvgSpotLen        Bases   BioProject     BioSample  \\\n",
       "0  SRR6507277        WGS         300  16133582100  PRJNA385855  SAMN08390922   \n",
       "1  SRR6507278        WGS         300  15874959000  PRJNA385855  SAMN08390923   \n",
       "2  SRR6507279        WGS         300  15069825300  PRJNA385855  SAMN08390924   \n",
       "3  SRR6507280        WGS         300  25807308000  PRJNA385855  SAMN08390925   \n",
       "4  SRR5720219        WGS         300   6713331000  PRJNA385855  SAMN07137016   \n",
       "\n",
       "                    BioSampleModel   bottle_id        Bytes Center Name  ...  \\\n",
       "0  MIMS.me,MIGS/MIMS/MIMARKS.water  2140200308   6618578156         MIT  ...   \n",
       "1  MIMS.me,MIGS/MIMS/MIMARKS.water  2160200304   6562862443         MIT  ...   \n",
       "2  MIMS.me,MIGS/MIMS/MIMARKS.water  1024800503   6265839401         MIT  ...   \n",
       "3  MIMS.me,MIGS/MIMS/MIMARKS.water  1025200510  10523504402         MIT  ...   \n",
       "4  MIMS.me,MIGS/MIMS/MIMARKS.water  1640201117   2811014041         MIT  ...   \n",
       "\n",
       "           lat_lon Library Name LibraryLayout LibrarySelection LibrarySource  \\\n",
       "0    22.75 N 158 W        S0627        PAIRED           RANDOM   METAGENOMIC   \n",
       "1    22.75 N 158 W        S0628        PAIRED           RANDOM   METAGENOMIC   \n",
       "2  31.66 N 64.16 W        S0629        PAIRED           RANDOM   METAGENOMIC   \n",
       "3  31.66 N 64.16 W        S0630        PAIRED           RANDOM   METAGENOMIC   \n",
       "4    22.75 N 158 W        S0519        PAIRED           RANDOM   METAGENOMIC   \n",
       "\n",
       "            Organism  Platform           ReleaseDate Sample Name  SRA Study  \n",
       "0  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0627  SRP109831  \n",
       "1  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0628  SRP109831  \n",
       "2  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0629  SRP109831  \n",
       "3  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0630  SRP109831  \n",
       "4  marine metagenome  ILLUMINA  2018-05-01T00:00:00Z       S0519  SRP109831  \n",
       "\n",
       "[5 rows x 36 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c53afba0-1ddb-4cb4-bfa9-00fb92ea6e03",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/jovyan/metagenomics_tutorial\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d3996d5-93a1-4a89-adc6-ad6a330e4307",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
